### PR TITLE
[ownership] update sil-lower-aggregate-instrs to support ossa

### PR DESCRIPTION
**Pull request by zoecarver**
This patch updates another pass to support ossa. 

I will probably need to make some changes to how the ossa is generated because I'm not super confident that I updated this correctly. Specifically, I was running into ownership lifetime errors when loading an address and passing that value to `struct_extract`/`tuple_extract` so I updated `emitLoweredDestroy` to handle addresses. I'm not sure that was the correct way to fix the issue, though. 

cc @gottesmm @atrick

### CODE REVIEW

**Comment by zoecarver** [comment]
I'll do the same thing I did in #29738 when I have the chance to get around to it.

**Comment by zoecarver** [comment]
**Pass after ownership elimination**
```
funcs changed: 898 transparent: 147 ossa: 147
```

**Pass before ownership elimination**
```
funcs changed: 898 transparent: 147 ossa: 1044
```

I was getting some weird results when the pass was in `addMandatoryOptPipeline` so I put it in `addPerfEarlyModulePassPipeline` instead (for the numbers above). I then got an error in `Mem2Reg`. It seemed like a legitimate bug in `Mem2Reg` so I fixed it and I didn't get any other errors. 

Besides that, the benchmarks / stdlib had no errors compiling while this pass was getting ossa.

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by theblixguy** [comment]
@swift-ci please test

**Comment by zoecarver** [comment]
Once `addPerfEarlyModulePassPipeline` supports ossa we can easily test both this and #30157 by passing them ossa _always_. It might be best to hold off on merging these until `addPerfEarlyModulePassPipeline` is updated so we can test them more thoroughly.

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by zoecarver** [comment]
@airspeedswift I don't that will do anything right now. Currently, this pass won't ever receive ossa (outside of transparent functions that will be removed anyway).

**Comment by airspeedswift** [comment]
yeah, mostly it's just in case something unintentionally regresses.

**Comment by zoecarver** [comment]
Fair enough.

File: lib/SIL/IR/TypeLowering.cpp [hunk 1]
```diff
@@ -817,15 +817,32 @@ namespace {
     void forEachNonTrivialChild(SILBuilder &B, SILLocation loc,
                                 SILValue aggValue,
                                 const T &operation) const {
+      // In ownership we can use destructure to lower destruction of aggValue.
```
Comment (zoecarver) [inline lib/SIL/IR/TypeLowering.cpp:820]: All changes in this file are from #30955.

**Comment by zoecarver** [comment]
@gottesmm I've updated, squashed, and simplified this patch. It should be easier to review now.

**Comment by zoecarver** [comment]
Closing in favor of #32885.

**Pull request closed without merging**